mantidproject/mantid
# MonteCarlo absorption calculation - reuse simulated tracks for different wavelengths

**Description of work.**

A switch has been added to the MCAbsorptionStrategy class that allows a pair
of before scatter\after scatter tracks to be used to calculate the attenuation
for a range of different lambda values. This will significantly improve the speed of the Monte
Carlo calculation. The Monte Carlo calculation currently sets the switch to
reuse the tracks without exposing this in the algorithm parameters - I've mainly
left the "old" approach in place in case any further wavelength dependent effects
are added to the absorption calculation eg beam spreading

In order to add the reuse while retaining the original behaviour (via the switch),
I had to reverse the order of the event and lambda loops. The previous code looped
over lambda with an inner loop of events. The event loop is now the outer loop, so
even with the switch set to not reuse tracks the results will change slightly if
the simulation isn't converged

The function MCInteractionVolume::calculateAbsorption has been split into two:
- track generated done in calculateBeforeAfterTrack
- absorption calculation done in calculateAbsorption

The relevant unit and system tests have been adjusted. In particular the tests in
MCInteractionVolumeTest have been split out so the track generation and absorption
calculation are tested separately

**To test:**

This change is mainly a performance improvement. So test involves running the Monte Carlo Absorption algorithm and checking the run time is less than before and the resulting attenuation plots are "similar" (not identical because the change means the random numbers used aren't exactly the same before\after). The following steps provide an example calculation on a real instrument:

Unzip [AbsorptionCorrectionSampling.zip](https://github.com/mantidproject/mantid/files/4123183/AbsorptionCorrectionSampling.zip) and place all the files in a directory that is specified Data Search Directories in "Manage User Directories". Run the .py script which does two things:
a) creates a workspace called "sample_env_mesh" containing
a mesh representation of the Pearl sample\environment with a single detector loaded. The script loads a mesh description of each environment component from a .stl file
b) runs the MonteCarloAbsorption algorithm on the sample_env_mesh workspace
to calculate an absorption correction and outputs the results into workspace
"sample_env_mesh_abscorr"

Plot the single spectrum in the sample_env_mesh_abscorr workspace (which is the absorption correction for the single detector) and check there's a valid absorption plot (~0.05-0.1 and decreasing with wavelength)

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.
